Working with APIs in TypeScript

cyberShaw opened the pull request:
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1880

srishilesh commented on articles/working-with-apis-in-typescript/index.md:
```diff
@@ -0,0 +1,296 @@
+    alt: TypeScript API example image
+---
+
+TypeScript was introduced by Microsoft as a superset of JavaScript. It takes the versatility of JavaScript and improves it with type definitions. TypeScript is used with a lot of popular frameworks like React.js and React Native to build enterprise applications.
```
Could you please elaborate shortly on `type definitions`?

srishilesh commented on articles/working-with-apis-in-typescript/index.md:
```diff
@@ -0,0 +1,296 @@
+
+### Prerequisites & Pre-Reading Material
+
+- A basic understanding of [HTTP Requests](https://www.section.io/engineering-education/http-requests-nodejs/) and [APIs](https://www.section.io/engineering-education/rest-api/).
```
You need not type the entire URL if you are referencing a Section article. Just the path `/engineering-education/http-requests-nodejs/` would suffice.

cyberShaw replied:
Understood :D

srishilesh commented on articles/working-with-apis-in-typescript/index.md:
```diff
@@ -0,0 +1,296 @@
+
+### Prerequisites & Pre-Reading Material
+
+- A basic understanding of [HTTP Requests](https://www.section.io/engineering-education/http-requests-nodejs/) and [APIs](https://www.section.io/engineering-education/rest-api/).
+- Fundamental concepts of [Node.js](https://www.section.io/engineering-education/history-of-nodejs/).
+- [TypeScript basics](https://www.section.io/engineering-education/a-friendly-beginner-guide-to-typescript/).
```
```suggestion
- A basic understanding of [HTTP Requests](/engineering-education/http-requests-nodejs/) and [APIs](/engineering-education/rest-api/).
- Fundamental concepts of [Node.js](/engineering-education/history-of-nodejs/).
- [TypeScript basics](/engineering-education/a-friendly-beginner-guide-to-typescript/).
```

srishilesh commented on articles/working-with-apis-in-typescript/index.md:
````diff
@@ -0,0 +1,296 @@
+npm init -y
+```
+
+Express is a framework built on Node.js. It is used to build backend servers and APIs. To learn more about Express, visit their [official docs](https://expressjs.com/en/4x/api.html).
````
`https://expressjs.com/en/4x/api.html` This link seems to be broken. It shows 404. Could you please add a working link?

cyberShaw replied:
![image](https://user-images.githubusercontent.com/34317373/114755909-c95e4680-9d77-11eb-8189-3c2b907ecba4.png)

This website seems to be working. Could you please check again?

srishilesh replied:
My bad. I guess I didn't copy the entire URL. It works fine.

cyberShaw replied:
No worries

srishilesh commented on articles/working-with-apis-in-typescript/index.md:
```diff
@@ -0,0 +1,296 @@
+
+Image: **Starting the API server**
+
+Now, open a separate terminal. We can test our API by the use of the `curl` command.
```
Could you add 1 - 2 lines explanation about `curl`? What's it used for?

srishilesh commented on articles/working-with-apis-in-typescript/index.md:
```diff
@@ -0,0 +1,296 @@
+  // The config function contains the routes
+  config() {
+    // The express.Request and express.Response types are defined in @types/express
+    this.app
+      .route("/api")
+      .get((req: express.Request, res: express.Response) => {
+        // The response is sent back
+        res.send(`GET Request successful!`);
+      });
+    // :id defines a query param
+    this.app
+      .route("/api/:id")
+      .get((req: express.Request, res: express.Response) => {
+        // The parameter is sent back as a response
+        res.send(
+          `GET Request successful. Parameter passed is: ${req.params.id}`
+        );
+      });
+    return this.app;
```
Apart from `GET` request, you could create 3 other APIs for `POST`, `PUT` and `DELETE` request, and explain about them. It would help the user understand how to handle them as well.

srishilesh reviewed (changes requested):
Hey @cyberShaw,

Awesome article! It's very beginner friendly.

I have suggested minor changes. Please let me know when they're done, we can move on to the tech review.

Thank you.

cyberShaw commented:
I'll add the `PUT`, `POST` and `DELETE` requests to the code today. I've made the other changes that you've requested. @srishilesh

srishilesh reviewed (changes requested):
Hey @cyberShaw,

I have done the tech review. Everything seems to work good. Great job!

Please let me know after you add the CRUD operation APIs, we can move this for the final review.

Thank you.

cyberShaw commented:
> Hey @cyberShaw,
> 
> I have done the tech review. Everything seems to work good. Great job!
> 
> Please let me know after you add the CRUD operation APIs, we can move this for the final review.
> 
> Thank you.

I'll do it by today :D

Thanks for reviewing my article, @srishilesh

hectorkambow commented:
plagiarism check done

srishilesh commented:
@cyberShaw

Any updates on the suggestions?

cyberShaw commented:
> @cyberShaw
> 
> Any updates on the suggestions?

Hey, @srishilesh 
Was a little busy this week. I'll make the suggested changes by today :D

cyberShaw commented:
@srishilesh I've added an example for POST. `PUT` and `DELETE` follows the exact same format so I've specified it in the article and also added a screenshot. Please check out the changes and let me know.

srishilesh reviewed (approved):
Hey @cyberShaw,

Thanks for making the requested changes. The article looks great now.

Moving to the final review.

Thank you.

hectorkambow commented:
👍 great work gents @cyberShaw 
cc @srishilesh

hectorkambow commented:
NOTE: Article is ready to be merged - we are making some changes to the main Repo - waiting for these to publish. 👍
